### PR TITLE
Replacing thread local with object pooling as a prevention for excess…

### PR DESCRIPTION
**Pull request by bilisie**
…ive leakage of weak references(caused by aggressive thread recycling in the beam framework).

### CODE REVIEW

**Comment by nevillelyh** [comment]
@kanterov @regadas PTAL

File: scio-test/src/test/scala/com/spotify/scio/coders/CaseClassSerializerTest.scala [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+class CaseClassSerializerTest extends PipelineSpec {
+
+  val kryo = () => new KryoAtomicCoder[Any](KryoOptions()).kryo().kryo
+
+  private def ccs[T:ClassTag](kryo: Kryo) = {
+    new CaseClassSerializer[T](kryo)
+  }
+
+  "KryoAtomicCoder" should "support Scala collections" in {
+    val instnace = kryo()
```
Comment (nevillelyh) [inline scio-test/src/test/scala/com/spotify/scio/coders/CaseClassSerializerTest.scala]: typo, `instance`

File: scio-test/src/test/scala/com/spotify/scio/coders/CaseClassSerializerTest.scala [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import com.esotericsoftware.kryo.io.{ByteBufferInput, FastOutput}
+import com.spotify.scio.testing.PipelineSpec
+
+import scala.reflect.ClassTag
+
+class CaseClassSerializerTest extends PipelineSpec {
+
+  val kryo = () => new KryoAtomicCoder[Any](KryoOptions()).kryo().kryo
+
+  private def ccs[T:ClassTag](kryo: Kryo) = {
```
Comment (nevillelyh) [inline scio-test/src/test/scala/com/spotify/scio/coders/CaseClassSerializerTest.scala]: This method is not used?

File: scio-test/src/test/scala/com/spotify/scio/coders/CaseClassSerializerTest.scala [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    new CaseClassSerializer[T](kryo)
+  }
+
+  "KryoAtomicCoder" should "support Scala collections" in {
+    val instnace = kryo()
+    instnace.addDefaultSerializer(classOf[Sample], new CaseClassSerializer[Sample](instnace))
+    instnace.newInstance(classOf[Sample]) should not be null
+    val output  = new FastOutput(0, 1024)
```
Comment (nevillelyh) [inline scio-test/src/test/scala/com/spotify/scio/coders/CaseClassSerializerTest.scala]: extra space before `=`

File: scio-core/src/main/scala/com/spotify/scio/coders/CaseClassSerializer.scala [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import com.esotericsoftware.kryo.Kryo
+import com.esotericsoftware.kryo.io.Input
+import com.esotericsoftware.kryo.serializers.FieldSerializer
+import com.spotify.scio.util.ScioUtil
+import org.objenesis.instantiator.ObjectInstantiator
+import org.objenesis.instantiator.sun.SunReflectionFactoryInstantiator
+
+import scala.reflect.ClassTag
+
+class CaseClassSerializer[T: ClassTag](kryo: Kryo)
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/coders/CaseClassSerializer.scala]: The user has to manually create this for each case class type right? Can you include a short example?

Comment (bilisie) [reply scio-core/src/main/scala/com/spotify/scio/coders/CaseClassSerializer.scala]: Will do.

File: scio-core/src/main/scala/com/spotify/scio/coders/KryoAtomicCoder.scala [hunk 5]
```diff
@@ -79,100 +80,91 @@ private[scio] class KryoAtomicCoder[T](private val options: KryoOptions) extends
...
-        new AlgebirdRegistrar()(k)
-        KryoRegistrarLoader.load(k)
-
-        val input = new InputChunked(options.bufferSize)
-        val output = new OutputChunked(options.bufferSize)
-
-        KryoState(k, input, output)
-      }
+  private val kryoPool: EmptyOnSerializeAtomicReference[ObjectPool[KryoState]]
+  = new EmptyOnSerializeAtomicReference[ObjectPool[KryoState]]() {
```
Comment (jto) [inline scio-core/src/main/scala/com/spotify/scio/coders/KryoAtomicCoder.scala]: Why do you need that intermediate `EmptyOnSerializeAtomicReference` here. I don't get the need for an `AtomicReference``

**Comment by jto** [comment]
CI seems to get stuck on this PR. Appart from that, LGTM.

File: scio-core/src/main/scala/com/spotify/scio/coders/CaseClassSerializer.scala [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+import com.esotericsoftware.kryo.io.Input
+import com.esotericsoftware.kryo.serializers.FieldSerializer
+import com.spotify.scio.util.ScioUtil
+import org.objenesis.instantiator.ObjectInstantiator
+import org.objenesis.instantiator.sun.SunReflectionFactoryInstantiator
+
+import scala.reflect.ClassTag
+
+/**
+  * Optimized serializer that can be used for case classes.
```
Comment (regadas) [inline scio-core/src/main/scala/com/spotify/scio/coders/CaseClassSerializer.scala]: we use javadoc style ... we have more code style info in the wiki.

File: scio-core/src/main/scala/com/spotify/scio/coders/ObjectPoolUtils.scala [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import org.apache.commons.pool2.impl.{DefaultPooledObject, GenericObjectPool,
+  GenericObjectPoolConfig}
+import org.apache.commons.pool2.{ObjectPool, PooledObject, PooledObjectFactory}
+import scala.util.Try
+
+object ObjectPoolUtils {
+
+
+  def createPool[T](factory: () => T) : ObjectPool[T] = {
+    new GenericObjectPool[T](new PooledObjectFactory[T] {
```
Comment (regadas) [inline scio-core/src/main/scala/com/spotify/scio/coders/ObjectPoolUtils.scala]: it seems that kryo was it's own pooling support. what do you think of giving it a try first and introduce commons-pool if doesn't work out? 

https://github.com/EsotericSoftware/kryo#pooling-kryo-instances

They provide an example and i think it's pretty straightforward to replace commons-pool.

**Comment by oskoi** [comment]
@bilisie when will you be able to correct the remarks? Is there anything I can do? Really looking forward to this PR

**Comment by nevillelyh** [comment]
@bilisie @oskoi please take a look at #1119 as well. It's revised based on this PR.

**Comment by oskoi** [comment]
@nevillelyh  Awesome!

**Comment by ravwojdyla** [comment]
Closing - moving to #1119

**Pull request closed without merging**